distributed/semaphore.py

```diff
@@ -138,37 +181,44 @@ def release(self, comm=None, name=None, client=None, identifier=None):
                 return
             if isinstance(name, list):
                 name = tuple(name)
-            if name in self.leases and identifier in self.leases[name]:
-                self._release_value(name, client, identifier)
+            if name in self.leases and lease_id in self.leases[name]:
+                self._release_value(name, lease_id)
             else:
-                raise ValueError(
+                logger.warning(
                     f"Tried to release semaphore but it was already released: "
-                    f"client={client}, name={name}, identifier={identifier}"
+                    f"name={name}, lease_id={lease_id}. This can happen if the semaphore timed out before."
                 )
 
-    def _release_value(self, name, client, identifier):
+    def _release_value(self, name, lease_id):
+        logger.info("Releasing %s for %s", lease_id, name)
         # Everything needs to be atomic here.
-        self.leases_per_client[client][name].remove(identifier)
-        self.leases[name].remove(identifier)
+        del self.leases[name][lease_id]
         self.events[name].set()
 
-    def _release_client(self, client):
-        semaphore_names = list(self.leases_per_client[client])
-        for name in semaphore_names:
-            ids = list(self.leases_per_client[client][name])
-            for _id in list(ids):
-                self._release_value(name=name, client=client, identifier=_id)
-        del self.leases_per_client[client]
-
     def _validate_leases(self):
         if not self._validation_running:
             self._validation_running = True
-            known_clients_with_leases = set(self.leases_per_client.keys())
-            scheduler_clients = set(self.scheduler.clients.keys())
-            for dead_client in known_clients_with_leases - scheduler_clients:
-                self._release_client(dead_client)
-            else:
-                self._validation_running = False
+            now = time()
+            semaphore_names = list(self.leases.keys())
+            for name in semaphore_names:
+                ids = list(self.leases[name])
+                logger.debug(
+                    "Validating leases for %s at time %s. Currently known %s",
+                    name,
+                    now,
+                    self.leases[name],
+                )
+                for _id in ids:
+                    time_since_refresh = now - self.leases[name][_id]
+                    if time_since_refresh > self.lease_timeout:
+                        logger.info(
+                            "Lease %s for %s timed out after %ss.",
+                            _id,
+                            name,
+                            time_since_refresh,
+                        )
+                        self._release_value(name=name, lease_id=_id)
+            self._validation_running = False
```

Comment:
shouldn't this be unintended once?

Reply:
No this is correct. Think of this as a poor-mans lock and the setting to False would be a release. If we'd unindent once we'd release every time the PC is called. This _lock_, however, is supposed to ensure that the body is only executed by one coroutine. While writing this I realise that the validation, by now, is entirely synchronous and this cannot happen at all. This guard can be removed.

Reply:
Sorry, I misread the indentation here, thought this was inside the loop